pypicloud/storage/azure_blob.py

```diff
@@ -0,0 +1,180 @@
+        path = self.get_path(package)
+
+        url_params = generate_blob_sas(
+            account_name=self.azure_storage_account_name,
+            container_name=self.azure_storage_container_name,
+            blob_name=path,
+            account_key=self.azure_storage_account_key,
+            permission=BlobSasPermissions(read=True),
+            expiry=datetime.now() + timedelta(seconds=self.expire_after),
+            protocol="https",
+        )
+
+        url = "{}/{}/{}?{}".format(
+            self.azure_storage_account_url,
+            self.azure_storage_container_name,
+            path,
+            url_params,
+        )
+        return url
+
+    def download_response(self, package):
+        return HTTPFound(location=self._generate_url(package))
+
+    def list(self, factory=Package):
+        # List does not return metadata :(
+        for blob_properties in self.container_client.list_blobs(
+            name_starts_with=self.path_prefix
+        ):
+            blob_client = self.container_client.get_blob_client(
+                blob=blob_properties.name
+            )
+            metadata = blob_client.get_blob_properties()
+
+            yield factory(
+                metadata.metadata["name"],
+                metadata.metadata["version"],
+                posixpath.basename(blob_properties.name),
+                blob_properties.last_modified,
+                path=blob_properties.name,
+                **Package.read_metadata(metadata.metadata)
+            )
+
+    def get_path(self, package):
+        """ Get the fully-qualified bucket path for a package """
+        if "path" not in package.data:
+            package.data["path"] = (
+                self.path_prefix + package.name + "/" + package.filename
+            )
+        return package.data["path"]
+
+    def upload(self, package, datastream):
+        path = self.get_path(package)
+
+        metadata = package.get_metadata()
+        metadata["name"] = package.name
+        metadata["version"] = package.version
+        self._normalize_metadata(metadata)
+
+        blob_client = self.container_client.get_blob_client(blob=path)
+        blob_client.upload_blob(data=datastream, metadata=metadata)
+
+    @staticmethod
+    def _normalize_metadata(metadata):
+        """Strip non-ASCII characters from metadata"""
+        for key, value in metadata.items():
+            if isinstance(value, six.string_types):
+                if isinstance(value, six.binary_type):
+                    value = value.decode("utf-8")
+                metadata[key] = "".join(
+                    c for c in unicodedata.normalize("NFKD", value) if ord(c) < 128
+                )
+
+    def delete(self, package):
+        path = self.get_path(package)
+        blob_client = self.container_client.get_blob_client(blob=path)
+        blob_client.delete_blob()
+
+    def check_health(self):
+        try:
+            list(self.container_client.list_blobs(name_starts_with=""))
```

Comment:
This looks like an expensive operation. Is there anything simpler we could do here? All this is used for is to check that the connection works and the auth credentials are valid.

Reply:
Good shout, will change it to looking for properties of a singular resource, as long as we get a response it'll be good.